controllers/openstackmachine_controller.go

```diff
@@ -292,11 +302,155 @@ func (r *OpenStackMachineReconciler) reconcileDelete(scope scope.Scope, cluster
+				conditions.MarkFalse(openStackMachine, infrav1.FloatingAddressFromPoolReadyCondition, infrav1.FloatingAddressFromPoolErrorReason, clusterv1.ConditionSeverityError, "Failed to reconcile floating IP claims: %v", err)
+				return nil, err
+			}
+			scope.Logger().Info("Created IPAddressClaim", "name", claim.Name)
+		} else {
+			conditions.MarkFalse(openStackMachine, infrav1.FloatingAddressFromPoolReadyCondition, infrav1.FloatingAddressFromPoolErrorReason, clusterv1.ConditionSeverityError, "Failed to reconcile floating IP claims: %v", err)
+			return nil, err
+		}
+	}
+
+	if claim.Status.AddressRef.Name == "" {
+		scope.Logger().Info("Waiting for IPAddressClaim to be allocated", "name", claim.Name)
+		openStackMachine.Status.FloatingAddressFromPoolReady = pointer.Bool(false)
+		conditions.MarkFalse(openStackMachine, infrav1.FloatingAddressFromPoolReadyCondition, infrav1.FloatingAddressFromPoolWaitingForIpamProviderReason, clusterv1.ConditionSeverityWarning, "")
+		return claim, nil
+	}
+	conditions.MarkTrue(openStackMachine, infrav1.FloatingAddressFromPoolReadyCondition)
+	return claim, nil
+}
+
+func (r *OpenStackMachineReconciler) associateIPAddressFromIPAddressClaim(ctx context.Context, scope scope.Scope, openStackMachine *infrav1.OpenStackMachine, openStackCluster *infrav1.OpenStackCluster, instanceStatus *compute.InstanceStatus, instanceNS *compute.InstanceNetworkStatus, claim *ipamv1.IPAddressClaim) error {
+	networkingService, err := networking.NewService(scope)
+	if err != nil {
+		return err
+	}
+	computeService, err := compute.NewService(scope)
+	if err != nil {
+		return err
+	}
+
+	address := &ipamv1.IPAddress{}
+	addressKey := client.ObjectKey{Namespace: openStackMachine.Namespace, Name: claim.Status.AddressRef.Name}
+
+	if err := r.Client.Get(ctx, addressKey, address); err != nil {
+		return err
+	}
+
+	instanceAddresses := instanceNS.Addresses()
+	for _, instanceAddress := range instanceAddresses {
+		if instanceAddress.Address == address.Spec.Address {
+			openStackMachine.Status.FloatingAddressFromPoolReady = pointer.Bool(true)
+			conditions.MarkTrue(openStackMachine, infrav1.FloatingAddressFromPoolReadyCondition)
+			return nil
+		}
+	}
+	fip, err := networkingService.GetFloatingIP(address.Spec.Address)
+	if err != nil {
+		return err
+	}
+	if fip == nil {
+		conditions.MarkFalse(openStackMachine, infrav1.FloatingAddressFromPoolReadyCondition, infrav1.FloatingAddressFromPoolErrorReason, clusterv1.ConditionSeverityError, "floating IP does not exist")
+		return fmt.Errorf("floating IP %q does not exist", address.Spec.Address)
+	}
+
+	port, err := computeService.GetManagementPort(openStackCluster, instanceStatus)
+	if err != nil {
+		return err
+	}
+
+	// Add finalizer to claim to prevent deletion until the floating IP is released
+	if controllerutil.AddFinalizer(claim, infrav1.IPClaimMachineFinalizer) {
+		if err := r.Client.Update(ctx, claim); err != nil {
+			return err
+		}
+	}
```

Comment:
Please can you add this on creation instead of here? It's also not safe to use the IP address above before we've added the finalizer.